src/components/permissions/rowLevelPolicyHelpers.ts

```diff
@@ -0,0 +1,332 @@
+  ]
+*/
+const generateRowLevelPolicies = (permissionRows: Array<PermissionRow>) => {
+  let policiesObject: { [index: string]: Array<string> } = {}
+
+  permissionRows.forEach((permissionRow: PermissionRow) => {
+    const { permissionPolicyRules, templatePermissionId } = permissionRow
+    if (!permissionPolicyRules) return
+
+    const permissionNameAbbreviation = getPermissionNameAbbreviation(permissionRow)
+    policiesObject = {
+      ...policiesObject,
+      [permissionNameAbbreviation]: compileRowLevelPolicies(
+        permissionNameAbbreviation,
+        permissionPolicyRules
+      ),
+    }
+    // It's possible to have permission policy linked to permission name without any templatePermissionIds
+    if (!templatePermissionId) return
+
+    const templatePermissionAbbreviation = getTemplatePermissionAbbreviation(permissionRow)
+    policiesObject = {
+      ...policiesObject,
+      [templatePermissionAbbreviation]: compileRowLevelPolicies(
+        templatePermissionAbbreviation,
+        permissionPolicyRules
+      ),
+    }
+  })
+  // Turn from object (of keys and arrays) to flattened array
+  const policies = Object.values(policiesObject).flat()
+
+  return policies
+}
+
+/* Generates row level policies, for one set of rules (for one PermissionRow)
+  in "pp2pn2tp2", {
+      application: {
+        view: {
+          template_id: 'jwtPermission_bigint_templateId',
+          user_id: 'jwtUserDetails_bigint_userId',
+        },
+      },
+    }
+  out [
+    CREATE POLICY "view_pp2pn2tp2" ON "application" FOR SELECT USING (jwt_get_boolean('pp2pn2tp2') = true and user_id = jwt_get_bigint('userId') AND template_id = jwt_get_bigint('pp2pn2tp2_templateId'))
+  ]
+*/
+const compileRowLevelPolicies = (permissionAbbreviation: string, permissionPolicyRules: object) => {
+  const policies: Array<string> = []
+
+  Object.entries(permissionPolicyRules).forEach(([tableName, rulesByType]) => {
+    let usingCondition = 'true'
+    const { view: viewRules } = rulesByType
+    // Need to figure out USING clause for UPDATE rule that defines WITH CHECK via rules (for UPDATE, see compileRowLevelPolicy for more details)
+    // So the UPDATE policy will use USING clause from view (select) rules
+    if (viewRules)
+      usingCondition = replacePlaceholders(
+        getSqlConditionFromJSON(viewRules),
+        permissionAbbreviation
+      )
+
+    Object.entries(rulesByType).forEach(([ruleType, rules]) => {
+      // replacePlaceholders would replace `jwtUserDetails_bigint_${something}` with `jwt_get_bigint_${something}` etc..
+      const condition = replacePlaceholders(getSqlConditionFromJSON(rules), permissionAbbreviation)
+
+      policies.push(
+        compileRowLevelPolicy(
+          tableName,
+          permissionAbbreviation,
+          ruleType,
+          usingCondition,
+          condition
+        )
+      )
+    })
+  })
```

Comment:
My personal preference is for `map` instead of `forEach` here

Reply:
I think the reason I used forEach and push, it to avoid extra syntax for destructuring.
i.e.
```TS
let allPolicies: Array<string> = []
// ...
const subsetPolicies = Object.entries(rulesByType).map(.....)
allPolicies = [...subsetPolicies, ...allPolicies]
```

Reply:
I think it's another of those stylistic preferences.

In the forEach case, there is no need for concat or restructuring, so my point is, why add it ?